backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt

```diff
@@ -605,34 +606,37 @@ private fun ObjCExportCodeGenerator.generateContinuationToCompletionConverter(
     }
 }
 
+private val ObjCExportBlockCodeGenerator.functionNClasses get() = context.ir.symbols.functionIrClassFactory.builtClasses.mapNotNull {
+    with(it.descriptor as FunctionClassDescriptor) {
+        if (functionKind == FunctionClassDescriptor.Kind.Function &&
+                arity < ObjCExportBlockCodeGenerator.functionTypeConvertorsArityLimit)
+            BuiltInFictitiousFunctionIrClassFactory.FunctionalInterface(it, arity)
+        else null
+    }
+}
```

Comment:
Most of this code has nothing to do with Obj-C export, and actually relies on some implementation details of functional types generation machinery of the compiler. What is the motivation behind moving it out from `BuiltInFictitiousFunctionIrClassFactory`?

Reply:
Rationale: `functionNClasses` is a helper that used by `ObjCExportBlockCodeGenerator` only. It is not a primary API of `BuiltInFictitiousFunctionIrClassFactory`, so as a local helper it belongs to usage scope. Implementation is a simple filter of public `builtClasses` by its public properties. As opposite, the filter depends on `functionTypeConvertorsArityLimit` which seems to be an implementation detail of exporting logic.

Another solution may be: make it an extension of `BuiltInFictitiousFunctionIrClassFactory` declared locally in the scope of `ObjCExportBlockCodeGenerator`. Is it what you have in your mind?

Reply:
> Rationale: `functionNClasses` is a helper that used by `ObjCExportBlockCodeGenerator` only.

It still uses implementation details of function types generation machinery (e.g. `FunctionClassDescriptor`), so I find it inseparable part of this machinery.

Also backend should avoid using descriptors. Obj-C export code generation is clearly a part of backend, while `BuiltInFictitiousFunctionIrClassFactory` is not.

> As opposite, the filter depends on functionTypeConvertorsArityLimit which seems to be an implementation detail of exporting logic.

It is quite easy to apply the filter separately from using implementation details.

> Another solution may be: make it an extension of BuiltInFictitiousFunctionIrClassFactory declared locally in the scope of ObjCExportBlockCodeGenerator. Is that what you have in your mind?

I don't think so.